File: Data Structures/c/queue/ArrayQueue.c
#include <stdio.h>

int size();
int isEmpty();
int isFull();
void peek();
void enqueue(int data);
void dequeue();
void print();

int maxCapacity = 100;
int tail = -1;
int Q[100];

int main() {
    enqueue(1);
    enqueue(2);
    enqueue(4);
    peek();
    print();
    dequeue();
    dequeue();
    print();
    return 0; 
}

int size() {
    return tail + 1;
}

int isEmpty() {
    return (size() == 0);
}

int isFull() {
    return (size() == maxCapacity);
}

void peek() {
    if (isEmpty()) {
        printf("Queue kosong\n");
    } else {
        printf("Yang paling depan adalah %d\n", Q[tail]);
    }
}

void enqueue(int data) {
    if (isFull()) {
        printf("Queue penuh\n");
    } else {
        printf("Menambahkan %d ke queue\n", data);
        Q[++tail] = data;
    }
}

void dequeue() {
    if (isEmpty()) {
        printf("Queue kosong\n");
    } else {
        printf("Data yang dihapus adalah %d\n", Q[0]);
        for (int i = 0; i < tail; i++) {
            Q[i] = Q[i+1];
        }
        tail--;
    }
}

void print() {
    for (int i = tail; i >= 0; i--) {
        printf("%d ", Q[i]);
    }
    printf("\n");
}

File: Data Structures/java/queue/array/ArrayQueue.java
package queue.array;

import queue.Queue;
import queue.QueueEmptyException;
import queue.QueueFullException;

public class ArrayQueue implements Queue {
    public static final int DEFAULT_CAPACITY = 100;
    private Object Data[];
    private int tail = -1; 
    private int capacity;

    //default constructor
    public ArrayQueue() {
        this(DEFAULT_CAPACITY);
    }

    public ArrayQueue(int capacity) {
        this.capacity = capacity;
        Data = new Object[capacity];
    }

    public int size() {
        return tail + 1;
    }

    public boolean isFull() {
        return (size() == capacity);
    }

    public boolean isEmpty() {
        return (size() == 0);
    }

    public void enqueue(Object elem) throws QueueFullException {
        if (isFull()) {
            throw new QueueFullException("Queue is full");
        }
        Data[++tail] = elem;
    }

    public Object dequeue() throws QueueEmptyException {
        if (isEmpty()) {
            throw new QueueEmptyException("Queue is empty");
        }
        Object elem = Data[0];
        for (int i = 0; i < tail; i++) {
            Data[i] = Data[i + 1];
        }
        tail--;
        return elem;
    }

    public Object peek() throws QueueEmptyException {
        if (isEmpty()) {
            throw new QueueEmptyException("Queue is empty");
        }
        return Data[0];
    }
}
